<h2 style = "color : Green">Simple Linear Regression/ Multiple Linear Regression </h2><br>

<b><i>Multivariate Regression is a supervised machine learning algorithm involving multiple data variables for analysis. It is used when we want to predict the value of a variable based on the value of two or more different variables

<b>We proceed to the model building part and here we may use multivariate linear regression algorithm, you may choose a perfect ratio for train test split, train the model with linear regression and perform a summary operation and list out all the different parameters of regression line fitted and point out the key statistics from summary

In [2]:
## importing the libraries


from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np

In [14]:
dataset = pd.read_csv('cleaned_advertising.csv')
dataset

Unnamed: 0     TV  Radio  Newspaper  Sales
0             0  230.1   37.8       69.2   22.1
1             1   44.5   39.3       45.1   10.4
2             2   17.2   45.9       69.3   12.0
3             3  151.5   41.3       58.5   16.5
4             4  180.8   10.8       58.4   17.9
..          ...    ...    ...        ...    ...
195         195   38.2    3.7       13.8    7.6
196         196   94.2    4.9        8.1   14.0
197         197  177.0    9.3        6.4   14.8
198         198  283.6   42.0       66.2   25.5
199         199  232.1    8.6        8.7   18.4

[200 rows x 5 columns]

<H3>DATA PREPROCESSING</H3>

In [4]:
### Define X&Y ###

x = dataset.iloc[:, :-1]
y = dataset.iloc[:, -1].values

<h3>Model Training</h3>

<h3 style = "color : red">Create and fit a liner regression model to the Data for the prediction</h3>

In [5]:
### Splitting the Dataset ###
x_train, x_test, y_train, y_test = train_test_split(x,y)

In [6]:
lin = LinearRegression(fit_intercept= True)
lin.fit(x_train, y_train)

LinearRegression()

In [7]:
LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

LinearRegression(normalize=False)

In [8]:
print(f"Linear coefficients : {lin.coef_}")
print(f"Intercept : {lin.intercept_}")

Linear coefficients : [0.05420799 0.10063365 0.00361576]
Intercept : 4.771045959404079


<h3>Model Prediction</h3>

In [9]:
y_pred = lin.predict(x_test)

<h3>Metrices</h3>

In [10]:
### RMSE ###

from sklearn import metrics

print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

Root Mean Squared Error: 1.443263375377067


In [11]:
### R-squared ###


from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.9215350327884629

<h3>Tuning</h3>

estimate the relevancy of the coefficients

Now that you have coefficients, how can you tell if they are relevant to predict your target?

The best way is to find the p-value. The p-value is used to quantify statistical significance; it allows to tell whether the null hypothesis is to be rejected or not.

<b>The null hypothesis?

For any modelling task, the hypothesis is that there is some correlation between the features and the target. The null hypothesis is therefore the opposite: there is no correlation between the features and the target.

So, finding the p-value for each coefficient will tell if the variable is statistically significant to predict the target. 
As a general rule of thumb, if the p-value is less than 0.05: there is a strong relationship between the variable and the target.

In [12]:
import statsmodels.api as sm

In [15]:
X2 = sm.add_constant(x)
est = sm.OLS(y, X2)
est2 = est.fit()
print(est2.summary())
     

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.903
Model:                            OLS   Adj. R-squared:                  0.901
Method:                 Least Squares   F-statistic:                     605.4
Date:                Tue, 25 Apr 2023   Prob (F-statistic):           8.13e-99
Time:                        10:30:18   Log-Likelihood:                -383.34
No. Observations:                 200   AIC:                             774.7
Df Residuals:                     196   BIC:                             787.9
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          4.6251      0.308     15.041      0.0

<h4>P value</h4>

Looking at all coefficients, we have a p-value that is very low (although it is probably not exactly 0) for TV and Radio. This means that there is a strong correlation between these coefficients and the target (Sales).

Of course, we notice that the third coefficient (the one for newspaper) has a large p-value 0.954. Therefore, ad spending on newspaper is not statistically significant. Removing that predictor would slightly reduce the R² value, but we might make better predictions.

<h4>R² value and Adj. R² value</h4>

Then, looking at the R² value and Adj. R² value, we have 0.9. Therefore, about 90% of the variability of sales is explained by the amount spent on all ads

<h4>F-statistic</h4>

This is a good indicator of whether there is a relationship between Y and X. The further our F-statistic is away from 1, the better our regression model.The F-statistic is 605.4. This is much greater than 1, and since our data set if fairly small (only 200 data points), it demonstrates that there is a strong relationship between ad spending and sales.

<h4>std error</h4>

The coefficient standard error, as mentioned above, measures how much our coefficient estimates vary from the actual average value of our response variable. In other words, it measures the accuracy of coefficient estimates. The closer our standard error is to zero, the better.

<h4>t-value</h4>

The coefficient t-value measures how far (in standard deviations) our coefficient estimate is from 0. A large t-value, relative to standard error, would provide evidence against the null hypothesis and indicate that a relationships exists between the predictor and response variables. Predictors with low t-statistics can be dropped. Ideally, the t-value should be greater than 1.96 for a p-value to be less than 0.05.

<h3>Evaluate the performance of the trained model on the testing data¶</h3>

In [ ]:
# Use the trained model to make predictions on new data
new_data = pd.DataFrame({'TV': [10000], 'Radio': [5000], 'Newspaper': [2000]})
prediction = lin.predict(new_data)
print("Prediction:", prediction)